globus_cli/commands/endpoint/server/delete.py

```diff
@@ -1,21 +1,90 @@
+from textwrap import dedent
 import click
 
-from globus_cli.parsing import common_options, endpoint_id_arg, server_id_arg
+from globus_cli.parsing import common_options, endpoint_id_arg
 from globus_cli.safeio import formatted_print, FORMAT_TEXT_RAW
 
-from globus_cli.services.transfer import get_client
+from globus_cli.services.transfer import get_client, get_endpoint_w_server_list
 
 
-@click.command('delete', help='Delete a server belonging to an endpoint')
+def _spec_to_matches(server_list, server_spec, mode):
+    """
+    mode is in {uri, hostname, hostname_port}
+
+    A list of matching server docs.
+    Should usually be 0 or 1 matches. Multiple matches are possible though.
+    """
+    assert mode in ('uri', 'hostname', 'hostname_port')
+
+    def match(server_doc):
+        if mode == 'hostname':
+            return server_spec == server_doc['hostname']
+        elif mode == 'hostname_port':
+            return (server_spec == '{}:{}'.format(
+                        server_doc['hostname'],
+                        server_doc['port']))
+        elif mode == 'uri':
+            return (server_spec == '{}://{}:{}'.format(
+                        server_doc['scheme'],
+                        server_doc['hostname'],
+                        server_doc['port']))
+        else:
+            raise NotImplementedError(
+                'Unreachable error! Something is very wrong.')
+
+    return [server_doc for server_doc in server_list
+            if match(server_doc)]
+
+
+def _detect_mode(server):
+    try:
+        int(server)
+        return 'id'
+    except ValueError:
+        pass
+
+    if '://' in server:
+        return 'uri'
+
+    if ':' in server:
+        return 'hostname_port'
+
+    return 'hostname'
+
+
+@click.command('delete', help="""\
+    Delete a server belonging to an endpoint
+
+    SERVER may be a server ID, HOSTNAME, HOSTNAME:PORT, or URI
+    ( SCHEME://HOSTNAME:PORT )
+    """)
 @common_options
 @endpoint_id_arg
-@server_id_arg
-def server_delete(endpoint_id, server_id):
+@click.argument('server')
+def server_delete(endpoint_id, server):
     """
     Executor for `globus endpoint server show`
     """
     client = get_client()
-    server_doc = client.delete_endpoint_server(endpoint_id, server_id)
 
-    formatted_print(server_doc, text_format=FORMAT_TEXT_RAW,
+    mode = _detect_mode(server)
+
+    if mode != 'id':
+        endpoint, server_list = get_endpoint_w_server_list(endpoint_id)
```

Comment:
You may want to raise a usage error here if the `server_list` returned by `get_endpoint_w_server_list` returns `'S3'` (otherwise `_spec_to_matches` will be operating on a string instead of the iterable response it is actually expecting):

```
~$ globus endpoint server delete cf9bcaa5-6d04-11e5-ba46-22000b92c6ec example.com
TypeError: string indices must be integers

program returned status code 1
```

Reply:
Adding to this, if you do decide to have a `'S3'` check here: you *could* go ahead and call `get_endpoint_w_server_list` regardless of the `mode` so that the user always gets consistent messaging:

```
~$ globus endpoint server delete 0f27d74c-1b41-11e8-b6eb-0ac6873fc732 123456
Globus CLI Error: A Transfer API Error Occurred.
HTTP status:      400
request_id:       UEaLoAYy5
code:             BadRequest
message:          That endpoint type doesn't have servers

program returned status code 1

~$ globus endpoint server delete 0f27d74c-1b41-11e8-b6eb-0ac6873fc732 example.com
Usage: globus endpoint server delete [OPTIONS] ENDPOINT_ID SERVER

Error: 0f27d74c-1b41-11e8-b6eb-0ac6873fc732 (My Test Share) is a share and does not have servers.

To see details of the share, use
    globus endpoint show 0f27d74c-1b41-11e8-b6eb-0ac6873fc732

To list the servers on the share's host endpoint, use
    globus endpoint server list ddb59aef-6d04-11e5-ba46-22000b92c6ec


program returned status code 2
```

Reply:
Playing with this more, I *really* dislike having the different error modes based on what you pass. Agree with your suggestion -- we'll always list, even if we don't need it.

I started to believe myself a little bit too much with the tests I wrote and forgot entirely to check on the S3 case (which the testsuite can't do because S3 endpoint creation is hosted CLI only). Will rectify along with the aforementioned fix.